onnx/defs/tensor/defs.cc

```diff
@@ -89,7 +89,98 @@ ONNX_OPERATOR_SET_SCHEMA(
             {"tensor(float16)", "tensor(float)", "tensor(double)"},
             "Constrain input and output types to float tensors.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          // Type Inference
           propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          // Shape Inference if target shape is in initializer
+          const TensorProto* targetShapeInitializer = ctx.getInputInitializer(1);
+          if (!targetShapeInitializer)  {
+            return;
+          }
+          // Make targetShape (0 -> same as originalShape, -1 -> inferred).
+          // The targetShape vector represents the specified shape for output.
+          std::vector<int64_t> targetShape;
+          if (targetShapeInitializer->has_raw_data())  {
+            const std::string& bytes = targetShapeInitializer->raw_data();
+            targetShape.insert(targetShape.end(), reinterpret_cast<const int64_t*>(bytes.c_str()), 
+                               reinterpret_cast<const int64_t*>(bytes.c_str() + bytes.size()));
+          } else {
+            const auto& data = targetShapeInitializer->int64_data();
+            targetShape.insert(targetShape.end(), data.begin(), data.end());
+          }
+
+          // Iterate through targetShape, adding dimensions in the outputShape TensorProto. 
+          // If the targertShape dimension is -1, we do not set the dimension value in this 
+          // iteration, but we record the Dimension. If targertShape dimension is 0, we attempt to infer the 
+          // dimension value. If it can be inferred, we set inferred dimension value; if not,
+          // we set the unresolvedZero flag. If targetShape dimension is positive,
+          // we set the dimension value in the outputShape. We track the product of 
+          // the dimensions we are setting outputShape in the outputProduct variable.
+          // The outputProduct will potentially be used for inferring a dimension marked -1.
+          auto* outputShape = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          TensorShapeProto::Dimension* negativeOneDim = nullptr;
+          const auto& dataInputTensorType = ctx.getInputType(0)->tensor_type();
+          bool unresolvedZero = false;
+          int64_t outputProduct = 1;
+          for (int i = 0; i < static_cast<int>(targetShape.size()); ++i)  {
```

Comment:
why use signed instead of unsigned for i here?

Reply:
Protobuf dim_size function returns an int, and I compare i < dataInputTensorType.shape().dim_size(). Also I think the dim function is supposed to take in an int